lasagne/layers/recurrent.py

```diff
@@ -803,8 +819,20 @@ def __init__(self, incoming, num_units,
         # inputs - the layer input, and the mask.  We will just provide the
         # layer input as incomings, unless a mask input was provided.
         incomings = [incoming]
+        self.mask_incoming_index = -1
+        self.hid_init_incoming_index = -1
+        self.cell_init_incoming_index = -1
         if mask_input is not None:
             incomings.append(mask_input)
+            self.mask_incoming_index = len(incomings)-1
+        if isinstance(hid_init, Layer):
+            incomings.append(hid_init)
+            self.hid_init_incoming_index = len(incomings)-1
+            learn_init = False
+        if isinstance(cell_init, Layer):
+            incomings.append(cell_init)
+            self.cell_init_incoming_index = len(incomings)-1
+            learn_init = False
```

Comment:
You shouldn't set `learn_init = False` here -- otherwise providing just one of `hid_init` or `cell_init` as a Layer will disable learning of the other. As it's only used within the `self.add_param()` call further below, the value of `learn_init` doesn't matter when the initial state is provided as a Layer anyway. So for consistence, you can also remove the `learn_init = False` assignment for the other recurrent layer classes.